### PR TITLE
Add UTC plugin

### PR DESCRIPTION
**Pull request by guisturdy**
```javascript
import utcPlugin from 'dayjs/plugin/utc'
dayjs.extend(utcPlugin)
// dayjs.extend(utcPlugin, { parseToLocal: Boolean })
```

### capability

```javascript

dayjs('2018-05-18T03:04:05+06:00')  // creating with utc

dayjs.utc()
dayjs.utc('2018-05-18T03:04:05+06:00') // (string | number | Date | Dayjs) support like dayjs()

dayjs().utc()
dayjs().local()

dayjs().isLocal()
dayjs().isUTC()

dayjs().utcOffset()   // get
dayjs().utcOffset(0)  // set

```

### TODO

- [x] doc
- [x] test cover

### CODE REVIEW

File: src/plugin/utc/util.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+function timezoneOffsetToString(tz) {
+  if (tz === 0) return 'Z'
+  const h = Math.floor(Math.abs(tz) / 60)
+  const m = Math.abs(tz) % 60
+  return `${tz > 0 ? '-' : '+'}${h < 10 ? '0' : ''}${h}${m < 10 ? '0' : ''}${m}`
```
Comment (Imperat) [inline src/plugin/utc/util.js]: Comment about ${h < 10 ? '0' : ''}${h}

Probably function padStart from utils of core code is better? :)
${padStart(h, 2, '0')}

Comment (guisturdy) [reply src/plugin/utc/util.js]: thanks, i replaced this method with `padZoneStr` from core code.

**Comment by Imperat** [comment]
In common I see that in this new plugin there is ability to parse timezone offset:
+06:00 or Z or -0530. 
According discussion in issue https://github.com/iamkun/dayjs/issues/143 do we want to add ability to parse timezone offset in core part of library? Just for have ability to parse next strings and get same results?
2018-05-18T03:04:05+06:00
2018-05-18T04:04:05+07:00

**Comment by guisturdy** [comment]
the core part already had the bility to parse timezone offset right now
```javascript
dayjs('2018-05-18T03:04:05+06:00').valueOf() // 1526591045000
dayjs('2018-05-18T04:04:05+07:00').valueOf() // 1526591045000
```
the difference after load this plugin is get data method (like `.year()` `.date()` `. format()`) will return the data based of the time zone what you set not the time zone of you local 
- not load plugin
```javascript
dayjs('2018-05-18T03:04:05+06:00').format() // 2018-05-18T05:04:05+08:00
dayjs('2018-05-18T04:04:05+07:00').format() // 2018-05-18T05:04:05+08:00
```
- load plugin
```javascript
dayjs.extend(utcPlugin)
dayjs('2018-05-18T03:04:05+06:00').format() // 2018-05-18T03:04:05+06:00
dayjs('2018-05-18T04:04:05+07:00').format() // 2018-05-18T04:04:05+07:00
```

**Comment by Imperat** [comment]
Got it. Thank you for clarification, @guisturdy!

**Comment by guisturdy** [comment]
#113

**Comment by iamkun** [comment]
Fantastic. Please allow me some time going through the code.

Plus, nice translation. Would you mind open a new PR with the translation of current Chinese Document only? So that I could merge it in advance.

**Comment by guisturdy** [comment]
@iamkun remove the part no belong to this plugin from Plugin.md and as an new PR #173

**Comment by iamkun** [comment]
@guisturdy  cool

**Comment by iamkun** [comment]
@guisturdy In order to keep the compiled file small and simple, we should keep away from `Array. reduce`  `Object.assign` and `{...}`

**Comment by guisturdy** [comment]
ok , i'll change this code to ES5 style

在 2018年5月22日，18:41，iamkun <notifications@github.com<mailto:notifications@github.com>> 写道：


@guisturdy<https://github.com/guisturdy> In order to keep the compiled file small and simple, we should keep away from Array. reduce Object.assign and {...}

**Comment by iamkun** [comment]
@guisturdy Thanks, ES6 style is OK and welcome, just not those modern APIs that are not fully supported by browsers. Those polyfills are too heavy to our project.

**Comment by guisturdy** [comment]
@iamkun finished

**Comment by hlehmann** [comment]
any update on this ?

**Comment by iamkun** [comment]
@hlehmann This is an import PR for our project. I'm trying to find some time go through this lot of code. And reviews are welcome.

**Comment by antony** [comment]
Would be great to have this in soon. Trying to use dayjs on a new project and `.startOf` is giving me yesterday at 11pm!

**Comment by RJiryes** [comment]
When will this be available? Thanks!

**Comment by guisturdy** [comment]
before this PR merged, hope [_**this**_](https://github.com/guisturdy/dayjs-plugin-utc) may help you
@RJiryes @antony

**Comment by prantlf** [comment]
Essentially, tis plugin retains the original time zone (offset) coming from the initialisation string. The date parts (year, month, ...) used in later formatting and other operation retain their original values. Am I right?

I am wondering, what would be a real-world use case for this. I mean no disrespect :-) Could you help me by explaining a bit, in what application and how you will use it?

Applications, which I am working on, need to display the time zone consistently, not to confuse the user. The current time zone parsing works well. Different date sources, potentially with different time zones in their strings, are parsed to standard `Date` object. It offers them in a single time zone - the local one - plus UTC support.

Some applications just work in the local time zone, because it is the most natural  for the users. No work needed, the `Date` object in Day.js supports it well.

Other applications let the users specify the time zone, which they wish to see and enter the date values. The `Date` object in Day.js does not support it. But this plugin does not support it either.

I opened a feature request for the full time zone support: #323.

**Comment by prantlf** [comment]
I browsed opened issues to help me understand the need for the "UTC mode". I found #74 and #249 particularly interesting.

It appears, that there is a misunderstanding about how Day.sj constructs `dayjs` instances. I have no problem, because I am aware, that Day.js behaves like the native `Date` - it assumes the local time zone, if the time part is omitted in the input string:

```js
dayjs("2018-10-28")
new Date("2018-10-28")
```

Both of the expressions above initialise the date instances to the following value in Central Europe: "2018-10-28T01:00+01:00". However, it is not obvious and people may think, that Day.js works in a "date-only mode", if they omit the the time part. If you know `Date` well, Day.js will not surprise you :-) But Day.js was created to make the the date handling easier. We could do better here.

This can be solved by two ways:

* Introduce a "date-only mode" in Day.js. Specify it during the construction, for example.
* Introduce an "UTC mode" in Day.js. The time will be always set to 00:00 and no conversion to the local time zone will be performed.

Although I feel, that the former would be clearer and better fitting for the use case, I admit, that the latter can be used to work with UTC dates as well and thus is more flexible.

When I was looking at this PR, I could not help myself not thinking, why would I need a new `UTCDate` class? The native `Date` stores the UTC value too and offers methods to inspect and manipulate it too. Is it not possible to stay with the `Date` and make the change simpler? Day.js stores the date and time parts in private variables $Y, $M etc. anyway. Most of the `dayjs` methods do not need to be modified.

I opened a new PR #326 with a simpler proposal. Not as a plugin, at first, because overriding `wrapper` and `parseDate` would mean copying practically all code to the plugin. The change is small, but similar to the introduction of the locale parameter.

**Comment by prantlf** [comment]
I suggest not to change behaviour of the constructor. If somebody includes the UTC plugin in an application, existing components already using `Day.js` should not get broken. I suggets making the "UTC mode" obvious by some parameter like `{ parseToLocal: Boolean }` or `{ assumeUTC: true }`.

Also, the `utcOffset` method should return the same sign, as `Date.prototype.getTimezoneOffset` for consistency. The method `month` returns also the month index instead of the month number to be consistent with `Date`.

**Comment by iamkun** [comment]
After careful consideration, we shouldn't change the behaviour of the original constructor. Thus, a plugin with `dayjs.utc` and `dayjs().utc()` is preferred and will be available soon.

**Pull request closed without merging**